twindb_backup/destination/ssh.py

```diff
@@ -124,64 +113,93 @@ def get_stream(self, path):
         else:
             return self._empty_status
 
     def _status_exists(self):
-        cmd = self._ssh_command + \
-              ["bash -c 'if test -s %s; "
+        cmd = ["bash -c 'if test -s %s; "
                "then echo exists; "
                "else echo not_exists; "
                "fi'" % self.status_path]
 
         try:
             LOG.debug('Running %r', cmd)
-            proc = Popen(cmd, stdout=PIPE, stderr=PIPE)
-            cout, cerr = proc.communicate()
-            if proc.returncode:
-                LOG.error('Failed to read backup status: %d: %s',
-                          proc.returncode,
-                          cerr)
+            stdout, _ = self._execute_commnand(cmd)
+            output = stdout.read()
+            if stdout.channel.recv_exit_status():
+                LOG.error('Failed to read backup status: %d',
+                          stdout.channel.recv_exit_status())
                 exit(1)
-            if cout.strip() == 'exists':
+            if output.strip() == 'exists':
                 return True
-            elif cout.strip() == 'not_exists':
+            elif output.strip() == 'not_exists':
                 return False
             else:
-                raise DestinationError('Unrecognized response: %s' % cout)
+                raise DestinationError('Unrecognized response: %s' % output)
 
         except OSError as err:
             LOG.error('Failed to run %s: %s', " ".join(cmd), err)
             exit(1)
+
     def share(self, url):
         super(Ssh, self).share(url)
+
+    def _execute_commnand(self, cmd):
+        """Execute ssh command"""
+        shell = SSHClient()
+        cmd_str = ' '.join(cmd)
+        try:
+            shell.connect(hostname=self.ssh_connect_info.host,
+                          key_filename=self.ssh_connect_info.key,
+                          port=self.ssh_connect_info.port,
+                          username=self.ssh_connect_info.user)
+            _, stdout, stderr = shell.exec_command(cmd_str)
+            return stdout, stderr
+        except (AuthenticationException, SSHException, socket.error) as err:
+            LOG.error("Failure execution %r : %s", cmd_str, err)
+            return -1
+        finally:
+            shell.close()
+
+    @contextmanager
+    def _get_remote_stdout(self, cmd):
+        """Get remote stdout handler"""
+        shell = SSHClient()
+        cmd_str = ' '.join(cmd)
+        try:
+            shell.connect(hostname=self.ssh_connect_info.host,
+                          key_filename=self.ssh_connect_info.key,
+                          port=self.ssh_connect_info.port,
+                          username=self.ssh_connect_info.user)
+            _, stdout, _ = shell.exec_command(cmd_str)
+            yield stdout
+        except (AuthenticationException, SSHException, socket.error) as err:
+            LOG.error("Failure execution %r : %s", cmd_str, err)
+            exit(1)
```

Comment:
Is `exit()` really necessary here? it would be nasty exit if it does so.
What about raising an exception?